File: sdk/python/src/safety_agent/utils/safe_url_fetcher.py
"""SSRF-safe remote URL fetching."""

import asyncio
from dataclasses import dataclass
import ipaddress
import socket
from urllib.parse import urljoin, urlparse

import aiohttp
from aiohttp.abc import AbstractResolver, ResolveResult

MAX_URL_LENGTH = 2048
MAX_REDIRECTS = 5
MAX_RESPONSE_BYTES = 25 * 1024 * 1024
REQUEST_TIMEOUT_SECONDS = 30.0


@dataclass
class PublicUrlResponse:
    final_url: str
    content_type: str
    data: bytes


def _validate_url_syntax(url: str) -> tuple[str, int]:
    if len(url) > MAX_URL_LENGTH:
        raise ValueError(
            f"Invalid URL: URL exceeds maximum length of {MAX_URL_LENGTH} characters"
        )

    try:
        parsed = urlparse(url)
        port = parsed.port or (443 if parsed.scheme == "https" else 80)
    except ValueError as error:
        raise ValueError(f"Invalid URL: malformed URL format - {error}") from error

    if parsed.scheme not in ("http", "https"):
        if parsed.scheme == "file":
            raise ValueError("Invalid URL: file:// protocol is not allowed")
        raise ValueError(
            f"Invalid URL: protocol must be http or https, got {parsed.scheme}"
        )
    if not parsed.hostname:
        raise ValueError("Invalid URL: hostname is required")
    if parsed.username or parsed.password:
        raise ValueError("Invalid URL: embedded credentials are not allowed")

    hostname = parsed.hostname.lower()
    if hostname in ("localhost", "localhost.localdomain", "local"):
        raise ValueError("Invalid URL: localhost access is not allowed")

    return parsed.hostname, port


async def _resolve_public_addresses(
    hostname: str, port: int
) -> list[tuple[str, int]]:
    try:
        direct_ip = ipaddress.ip_address(hostname)
        addresses = [
            (
                str(direct_ip),
                socket.AF_INET6 if direct_ip.version == 6 else socket.AF_INET,
            )
        ]
    except ValueError:
        try:
            results = await asyncio.get_running_loop().getaddrinfo(
                hostname,
                port,
                family=socket.AF_UNSPEC,
                type=socket.SOCK_STREAM,
            )
        except (socket.gaierror, socket.herror) as error:
            raise ValueError(
                "Invalid URL: hostname could not be resolved safely"
            ) from error

        addresses = []
        seen: set[tuple[str, int]] = set()
        for family, _type, _proto, _canonname, sockaddr in results:
            entry = (sockaddr[0], family)
            if entry not in seen:
                seen.add(entry)
                addresses.append(entry)

    if not addresses or any(
        not ipaddress.ip_address(address).is_global
        for address, _family in addresses
    ):
        raise ValueError("Invalid URL: private/internal IP addresses are not allowed")

    return addresses


class _PinnedResolver(AbstractResolver):
    def __init__(self, hostname: str, addresses: list[tuple[str, int]]) -> None:
        self._hostname = hostname.lower()
        self._addresses = addresses

    async def resolve(
        self,
        host: str,
        port: int = 0,
        family: socket.AddressFamily = socket.AF_INET,
    ) -> list[ResolveResult]:
        if host.lower() != self._hostname:
            raise OSError("Attempted to resolve an unvalidated hostname")

        return [
            ResolveResult(
                hostname=host,
                host=address,
                port=port,
                family=address_family,
                proto=socket.IPPROTO_TCP,
                flags=socket.AI_NUMERICHOST,
            )
            for address, address_family in self._addresses
            if family in (socket.AF_UNSPEC, address_family)
        ]

    async def close(self) -> None:
        return None


async def fetch_public_url(url: str) -> PublicUrlResponse:
    current_url = url
    timeout = aiohttp.ClientTimeout(total=REQUEST_TIMEOUT_SECONDS)

    for redirect_count in range(MAX_REDIRECTS + 1):
        hostname, port = _validate_url_syntax(current_url)
        addresses = await _resolve_public_addresses(hostname, port)
        connector = aiohttp.TCPConnector(
            resolver=_PinnedResolver(hostname, addresses),
            use_dns_cache=False,
            family=socket.AF_UNSPEC,
        )

        async with aiohttp.ClientSession(
            connector=connector,
            timeout=timeout,
            trust_env=False,
        ) as client:
            async with client.get(current_url, allow_redirects=False) as response:
                if 300 <= response.status < 400:
                    location = response.headers.get("location")
                    if not location:
                        raise RuntimeError(
                            "Failed to fetch URL: redirect is missing Location"
                        )
                    if redirect_count == MAX_REDIRECTS:
                        raise RuntimeError(
                            f"Failed to fetch URL: exceeded {MAX_REDIRECTS} redirects"
                        )
                    current_url = urljoin(current_url, location)
                    continue

                if response.status != 200:
                    raise RuntimeError(
                        f"Failed to fetch URL: {response.status} {response.reason}"
                    )
                if (
                    response.content_length is not None
                    and response.content_length > MAX_RESPONSE_BYTES
                ):
                    raise RuntimeError(
                        f"Failed to fetch URL: response exceeds {MAX_RESPONSE_BYTES} bytes"
                    )

                body = bytearray()
                async for chunk in response.content.iter_chunked(64 * 1024):
                    body.extend(chunk)
                    if len(body) > MAX_RESPONSE_BYTES:
                        raise RuntimeError(
                            f"Failed to fetch URL: response exceeds {MAX_RESPONSE_BYTES} bytes"
                        )

                return PublicUrlResponse(
                    final_url=current_url,
                    content_type=(
                        response.headers.get("content-type", "")
                        .split(";")[0]
                        .strip()
                    ),
                    data=bytes(body),
                )

    raise RuntimeError(f"Failed to fetch URL: exceeded {MAX_REDIRECTS} redirects")

File: sdk/typescript/src/utils/safe-url-fetcher.ts
import { lookup } from "node:dns/promises";
import ipaddr from "ipaddr.js";
import { Agent, fetch as undiciFetch } from "undici";

const MAX_URL_LENGTH = 2048;
const MAX_REDIRECTS = 5;
const MAX_RESPONSE_BYTES = 25 * 1024 * 1024;
const REQUEST_TIMEOUT_MS = 30_000;

interface ResolvedAddress {
  address: string;
  family: 4 | 6;
}

interface ValidatedUrl {
  parsed: URL;
  addresses: ResolvedAddress[];
}

export interface PublicUrlResponse {
  finalUrl: string;
  contentType: string;
  data: Uint8Array;
}

function isPublicIpAddress(ip: string): boolean {
  try {
    const address = ipaddr.process(ip);
    if (address.kind() === "ipv6") {
      const ipv6 = address as ipaddr.IPv6;
      if (ipv6.isIPv4MappedAddress()) {
        return ipv6.toIPv4Address().range() === "unicast";
      }
    }
    return address.range() === "unicast";
  } catch {
    return false;
  }
}

async function resolvePublicAddresses(
  hostname: string
): Promise<ResolvedAddress[]> {
  const cleanHostname = hostname.replace(/^\[|\]$/g, "");
  let addresses: ResolvedAddress[];

  try {
    if (ipaddr.isValid(cleanHostname)) {
      const address = ipaddr.process(cleanHostname);
      addresses = [
        {
          address: address.toString(),
          family: address.kind() === "ipv4" ? 4 : 6,
        },
      ];
    } else {
      const results = await lookup(cleanHostname, {
        all: true,
        verbatim: true,
      });
      addresses = results.map(({ address, family }) => ({
        address,
        family: family as 4 | 6,
      }));
    }
  } catch {
    throw new Error("Invalid URL: hostname could not be resolved safely");
  }

  if (
    addresses.length === 0 ||
    addresses.some(({ address }) => !isPublicIpAddress(address))
  ) {
    throw new Error(
      "Invalid URL: private/internal IP addresses are not allowed"
    );
  }

  return addresses;
}

async function validateUrl(url: string): Promise<ValidatedUrl> {
  if (url.length > MAX_URL_LENGTH) {
    throw new Error(
      `Invalid URL: URL exceeds maximum length of ${MAX_URL_LENGTH} characters`
    );
  }

  let parsed: URL;
  try {
    parsed = new URL(url);
  } catch (error) {
    throw new Error(`Invalid URL: malformed URL format - ${String(error)}`);
  }

  if (parsed.protocol !== "http:" && parsed.protocol !== "https:") {
    if (parsed.protocol === "file:") {
      throw new Error("Invalid URL: file:// protocol is not allowed");
    }
    throw new Error(
      `Invalid URL: protocol must be http or https, got ${parsed.protocol}`
    );
  }

  if (!parsed.hostname) {
    throw new Error("Invalid URL: hostname is required");
  }
  if (parsed.username || parsed.password) {
    throw new Error("Invalid URL: embedded credentials are not allowed");
  }

  const hostname = parsed.hostname.toLowerCase();
  if (
    hostname === "localhost" ||
    hostname === "localhost.localdomain" ||
    hostname === "local" ||
    hostname === "127.0.0.1" ||
    hostname === "::1" ||
    hostname.startsWith("127.")
  ) {
    throw new Error("Invalid URL: localhost access is not allowed");
  }

  return {
    parsed,
    addresses: await resolvePublicAddresses(parsed.hostname),
  };
}

function createPinnedAgent(addresses: ResolvedAddress[]): Agent {
  let nextAddress = 0;
  return new Agent({
    connect: {
      lookup: (_hostname, options, callback) => {
        if (options.all) {
          callback(null, addresses);
          return;
        }

        const selected = addresses[nextAddress % addresses.length];
        nextAddress += 1;
        callback(null, selected.address, selected.family);
      },
      timeout: REQUEST_TIMEOUT_MS,
    },
    headersTimeout: REQUEST_TIMEOUT_MS,
    bodyTimeout: REQUEST_TIMEOUT_MS,
    maxResponseSize: MAX_RESPONSE_BYTES,
  });
}

async function readResponseBody(
  response: Awaited<ReturnType<typeof undiciFetch>>
): Promise<Uint8Array> {
  const declaredLength = Number(response.headers.get("content-length"));
  if (
    Number.isFinite(declaredLength) &&
    declaredLength > MAX_RESPONSE_BYTES
  ) {
    throw new Error(
      `Failed to fetch URL: response exceeds ${MAX_RESPONSE_BYTES} bytes`
    );
  }

  if (!response.body) {
    return new Uint8Array();
  }

  const reader = response.body.getReader();
  const chunks: Uint8Array[] = [];
  let totalLength = 0;

  while (true) {
    const { done, value } = await reader.read();
    if (done) break;

    totalLength += value.byteLength;
    if (totalLength > MAX_RESPONSE_BYTES) {
      await reader.cancel();
      throw new Error(
        `Failed to fetch URL: response exceeds ${MAX_RESPONSE_BYTES} bytes`
      );
    }
    chunks.push(value);
  }

  const result = new Uint8Array(totalLength);
  let offset = 0;
  for (const chunk of chunks) {
    result.set(chunk, offset);
    offset += chunk.byteLength;
  }
  return result;
}

export async function fetchPublicUrl(url: string): Promise<PublicUrlResponse> {
  let currentUrl = url;

  for (let redirectCount = 0; redirectCount <= MAX_REDIRECTS; redirectCount++) {
    const validated = await validateUrl(currentUrl);
    const agent = createPinnedAgent(validated.addresses);

    try {
      const response = await undiciFetch(validated.parsed, {
        dispatcher: agent,
        redirect: "manual",
        signal: AbortSignal.timeout(REQUEST_TIMEOUT_MS),
      });

      if (response.status >= 300 && response.status < 400) {
        const location = response.headers.get("location");
        await response.body?.cancel();
        if (!location) {
          throw new Error("Failed to fetch URL: redirect is missing Location");
        }
        if (redirectCount === MAX_REDIRECTS) {
          throw new Error(
            `Failed to fetch URL: exceeded ${MAX_REDIRECTS} redirects`
          );
        }
        currentUrl = new URL(location, validated.parsed).toString();
        continue;
      }

      if (!response.ok) {
        await response.body?.cancel();
        throw new Error(
          `Failed to fetch URL: ${response.status} ${response.statusText}`
        );
      }

      return {
        finalUrl: currentUrl,
        contentType:
          response.headers.get("content-type")?.split(";")[0].trim() || "",
        data: await readResponseBody(response),
      };
    } finally {
      await agent.close();
    }
  }

  throw new Error(`Failed to fetch URL: exceeded ${MAX_REDIRECTS} redirects`);
}
